[react-form] include `dynamicLists` values in `onSubmit` type (#1865)
Shopify/quilt: packages/react-form/README.md

#### `useDynamicList()`

> A custom hook that adds on `useList` the ability to dynamically add, move and remove list tems.

Reviewer: ```suggestion
A custom hook that adds on `useList` the ability to dynamically add, move and remove list items.
```

What is the meaning of `moving a list` item?

##### Using `useDynamicList`

Let's simulate a user interface that allows you add as many credit cards as you wish.

> You can initialize `useDynamicList` the following way

Reviewer 2: ```suggestion
You can initialize `useDynamicList` the following way:
```

```tsx
interface Card {
  id: string;
  cardNumber: string;
  cvv: string;
}

const emptyCardFactory = (): Card => ({

>   id: new Date.now().toString(),

Reviewer 2: ```suggestion
  id: Date.now().toString(),
```

  cardNumber: '',
  cvv: '',
});

const {fields, addItem, removeItem, moveItem, reset, dirty} = useDynamicList(

>   [],

Author: True, that's why on L701 we have a dedicated example for that
> You can choose to initialize the list with no card (same as above) or with existing cards coming from your database for example.

  emptyCardFactory,
);
```

> You can also have a factory that produces multiple cards such as

Reviewer 2: ```suggestion
You can also have a factory that produces multiple cards such as:
```

```tsx
function generateId() {

>   return new Date.now().toString();

Reviewer 2: ```suggestion
  return Date.now().toString();
```

}

const emptyCardFactory = (): Card[] => {
  return [
    {id: generateId(), cardNumber: '', cvv: ''},
    {id: generateId(), cardNumber: '', cvv: ''},
  ];
};

const {fields, addItem, removeItem, moveItem, reset, dirty} = useDynamicList(
  [],
  emptyCardFactory,
);
```

> AddItem can accept an argument which will be passed to the factory. In this case, the factory could look like

Reviewer 2: ```suggestion
AddItem can accept an argument which will be passed to the factory. In this case, the factory could look like:
```

>         value={field.cvv.value}

Author: There is an example about that for `useField`, I think it's good to show what a field is made of.

        onChange={field.cvv.onChange}
      />
      <div>
        <Button onClick={() => removeItem(index)}>Remove</Button>

>       </div>

Author: I removed the `div` to reduce the density of the component tree.

      </Button>
      <Button
        disabled={index === fields.length - 1}
        onClick={() => moveItem(index, index + 1)}
      >
        Move Item Down
      </Button>
      <hr />
    </div>
  ))}
  <Button onClick={() => addItem()}>Add Card</Button>
  <Button disabled={!dirty} onClick={reset}>
    Reset
  </Button>
</FormLayout>
```

##### How to use it with `useForm`

You can use `useDynamicList` with `useForm` in two ways:

You can directly pass one or several dynamic lists within `fields`. But note that the `reset` function returned by `useForm` won't reset the dynamic list to its initial value. It will only reset the value of each field for each item. So for example if you remove an item, clicking on reset won't re-add it. Plus the adding or removing a field won't make the form dirty. The form will only be dirty if you edit the value of one item's field.

```tsx
const form = useForm({
  fields: {

>     customerCards: fields, // the fields destructured from useDynamicList

Reviewer: I would add the `useDynamicList` detructuring above, it would be easier to understand what we mean here.

  },
  onSubmit: async fieldValues => {
    console.log(fieldValues);
    return submitSuccess();
  },
});
```

> Or you can pass your dynamic lists within `dynamicLists`. This way the form `reset` and `dirty` will work as expected. Note that you still need to have one "classic field", `fields` cannot be optionnal at the moment. If you use TypeScript, you might want to pass generic types to `useForm` in order to make sure TypeScript will infer the `fieldValues` types correctly within `onSubmit`.

Reviewer: ```suggestion
Or you can pass your dynamic lists within `dynamicLists`. This way the form `reset` and `dirty` will work as expected. Note that you still need to have one "classic field", `fields` cannot be optional at the moment. If you use TypeScript, you might want to pass generic types to `useForm` in order to make sure TypeScript will infer the `fieldValues` types correctly within `onSubmit`.
```

    console.log(fieldValues);
    return {status: 'success'};
    return submitSuccess();
  },
});
```

you can directly pass a dynamic list within `fields` but note that the `reset` function returned by `useForm` will only reset each field, plus the `dirty` state will only reflect dirty fields within each item.
Or you can pass your dynamic lists within `dynamicLists`. This way the form `reset` and `dirty` will work as expected. Note that you still need to have one "classic field", `fields` cannot be optionnal at the moment. If you use TypeScript, you might want to pass generic types to `useForm` in order to make sure TypeScript will infer the `fieldValues` types correctly within `onSubmit`.

```tsx
const {submit, dirty, submitting, reset} = useForm({
  fields: {
    fields,  // the fields returned from useDynamicList.
const fields = {
  title: useField(''),
};

const customerCards = useDynamicList<Card>([], emptyCardFactory);

const form = useForm<typeof fields, typeof customerCards>({
  fields,
  dynamicLists: {
    customerCards,
  },
  onSubmit: async fieldValues => {
    console.log(fieldValues);
    return {status: 'success'};
    return submitSuccess();
  },
});
```

> A code sandbox of `useDynamicList` in action can be seen [here](https://codesandbox.io/s/hungry-rubin-exrkz?fontsize=14&hidenavigation=1&theme=dark)

Reviewer 3: I am not sure we should remove the sandbox as it helps people visualize how the dynamic list works

Author: If you think something i missing in the README then we can add it tho

        value={field.cvv.value}
        onChange={field.cvv.onChange}
      />
      <Button onClick={() => removeItem(index)}>Remove</Button>
      <Button disabled={index === 0} onClick={() => moveItem(index, index - 1)}>
        Move Item Up
      </Button>
      <Button
        disabled={index === fields.length - 1}
        onClick={() => moveItem(index, index + 1)}
      >
        Move Item Down
      </Button>
      <hr />
    </div>
  ))}
  <Button onClick={() => addItem()}>Add Card</Button>
  <Button disabled={!dirty} onClick={reset}>
    Reset
  </Button>
</FormLayout>
```

##### How to use it with `useForm`

You can use `useDynamicList` with `useForm` in two ways:

> You can directly pass one or several dynamic lists within `fields`. But note that the `reset` function returned by `useForm` won't reset the dynamic list to its initial value. It will only reset the value of each field for each item. So for example if you remove an item, clicking on reset won't re-add it. Plus the adding or removing a field won't make the form dirty. The form will only be dirty if you edit the value of one item's field.

Reviewer 2: ```suggestion
You can directly pass one or several dynamic lists within `fields`. But note that the `reset` function returned by `useForm` won't reset the dynamic list to its initial value. It will only reset the value of each field for each item. So, if you remove an item, clicking on reset won't re-add it. Plus, adding or removing a field won't make the form dirty. The form will only be dirty if you edit the value of an item's field.
```

```tsx
const {fields} = useDynamicList<Card>([], emptyCardFactory);

const form = useForm({
  fields: {
    customerCards: fields,
  },
  onSubmit: async fieldValues => {
    console.log(fieldValues);
    return submitSuccess();
  },
});
```

> Or you can pass your dynamic lists within `dynamicLists`. This way the form `reset` and `dirty` will work as expected. Note that you still need to have one "classic field", `fields` cannot be optional at the moment. If you use TypeScript, you might want to pass generic types to `useForm` in order to make sure TypeScript will infer the `fieldValues` types correctly within `onSubmit`.

Reviewer 2: ```suggestion
Or you can pass your dynamic lists within `dynamicLists`. This way the form `reset` and `dirty` will work as expected. Note that you still need to have one "classic field", `fields` cannot be optional at the moment. If you use TypeScript, you might want to pass generic types to `useForm` to make sure TypeScript infers the `fieldValues` types correctly within `onSubmit`.
```

```tsx
const fields = {
  title: useField(''),
};

const customerCards = useDynamicList<Card>([], emptyCardFactory);

const form = useForm<typeof fields, typeof customerCards>({
  fields,
  dynamicLists: {
    customerCards,
  },

>   onSubmit: async fieldValues => {

Reviewer 2: ```suggestion
  onSubmit: async (fieldValues) => {
```

- `config`, An object has the following fields:

  - `fields`, A dictionary of `Field` objects, dictionaries of `Field` objects, and lists of dictionaries of `Field` objects. Generally, you'll want these to be generated by the other hooks in this package, either `useField` or `useList`. This will be returned back out as the `fields` property of the return value.

>   - `useDynamicList`, optionnal dictionaries of `DynamicList`.

Reviewer: ```suggestion
  - `useDynamicList`, optional dictionaries of `DynamicList`.
```
